src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```diff
@@ -2,44 +2,107 @@
+    /// If this cmdlet errors then it throws a non-terminating error.
     /// </summary>
     [Cmdlet(VerbsCommon.New, "TemporaryFile", SupportsShouldProcess = true, HelpUri = "https://go.microsoft.com/fwlink/?LinkId=526726")]
     [OutputType(typeof(System.IO.FileInfo))]
     public class NewTemporaryFileCommand : Cmdlet
     {
         /// <summary>
-        /// returns a TemporaryFile
+        /// Specify a different file extension other than the default one, which is '.tmp'. The period in this parameter is optional.
+        /// </summary>
+        [Parameter(Position = 0)]
+        [ValidateNotNullOrEmpty]
+        public string Extension { get; set; } = ".tmp";
+
+        /// <summary>
+        /// Use a Guid as a file name to allow for more than 65,535 unique file names in the temporary folder.
+        /// </summary>
+        [Parameter()]
+        public SwitchParameter GuidBasedName  { get; set; }
+
+        /// <summary>
+        /// Returns a TemporaryFile.
         /// </summary>
         protected override void EndProcessing()
         {
+            // Check for invalid characters in extension
+            if (!string.IsNullOrEmpty(Extension) && Extension.IndexOfAny(Path.GetInvalidFileNameChars()) != -1)
+            {
+                WriteError(
+                    new ErrorRecord(
+                        new ArgumentException(
+                                StringUtil.Format(UtilityCommonStrings.InvalidCharacterInParameter,
+                                    nameof(Extension), Extension)),
+                        "NewTemporaryInvalidArgument",
+                        ErrorCategory.InvalidArgument,
+                        Extension));
+                return;
+            }
+
             string filePath = null;
-            string tempPath = System.Environment.GetEnvironmentVariable("TEMP");
+            string tempPath = Path.GetTempPath();
+
             if (ShouldProcess(tempPath))
             {
-                try
+                int attempts = 0;
+                bool creationOfFileSuccessful = false;
+
+                // In case the random temporary file already exists, retry 
+                while (attempts++ < 10 && !creationOfFileSuccessful)
                 {
-                    filePath = Path.GetTempFileName();
+                    try
+                    {
+                        string fileName;
+                        if (GuidBasedName.IsPresent)
+                        {
+                            fileName = Guid.NewGuid().ToString();
+                        }
+                        else
+                        {
+                            fileName = Path.GetRandomFileName();
+                        }
+                        fileName = Path.ChangeExtension(fileName, Extension);
+                        filePath = Path.Combine(tempPath, fileName);
+                        using (new FileStream(filePath, FileMode.CreateNew)) { }
```

Comment:
A comment would be good here - explaining why we create the file and do not remove it, similar to `Path.GetTempFileName`.